File: src/Bot.ts
/*
 * Notify
 * Copyright (c) 2021 BanDev
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { Client, Intents, VoiceState } from "discord.js";
import { Command } from "./Command";

export class Bot {

    client: Client;
    token: string;
    intents: Intents[];
    commands: Command[];
    ready: (client: Client) => void;

    constructor(token: string, intents: Intents[], commands: Command[]) {
        this.token = token;
        this.intents = intents;
        this.commands = commands;
    }

    public listen(ready: (client: Client) => void) {
        this.client = new Client({ intents: this.intents });
        this.ready = ready;
        this.client.login(this.token);
        this.onInteraction();
        this.onMessage();
        this.onReady();
    }

    public vcUpdate(respond: (before: VoiceState, after: VoiceState) => void) {
        this.client.on("voiceStateUpdate", 
        (before: VoiceState, after: VoiceState) => 
        {respond(before, after);}) 
    }

    /**
     * When an interaction is created, try to 
     * match the name of the command to a command
     * passed to the commands array of this class.
     */

    private onInteraction() {
        this.client.on("interactionCreate", async interaction => {
            if (!interaction.isCommand()) return;
            let command = this.commands.find(it => it.name === interaction.commandName);
            await command.run(interaction, null);
        });
    }

    private onReady() {
        this.client.on("ready", client => {
            client.user.setActivity('bandev.uk/notify', { type: 'PLAYING' });
            this.ready(client);
        });
    }

    private onMessage() {
        this.client.on("messageCreate", async message => {
            console.log(message)
        });
    }

}

File: src/main/kotlin/Bot.kt
/**

Notify
Copyright (C) 2022  BanDev

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

 */

import dev.kord.core.Kord
import dev.kord.core.event.interaction.ChatInputCommandInteractionCreateEvent
import dev.kord.core.event.message.MessageCreateEvent
import dev.kord.core.event.user.VoiceStateUpdateEvent
import dev.kord.core.on
import dev.kord.gateway.Intent
import dev.kord.gateway.Intents
import interactions.InteractionCommand
import io.klogging.Klogging
import vcStateChange.StateChange

class Bot: Klogging {

    lateinit var client: Kord

    suspend fun init(): Bot {
        client = Kord(Config.discordApiKey)
        return this
    }

    suspend fun listenInteractions(vararg interactions: InteractionCommand) {
        client.on<ChatInputCommandInteractionCreateEvent> {
            val chosen = arrayOf(*interactions).find { interactionCommand -> interactionCommand.name == interaction.name }

            if (chosen != null) chosen.call(interaction)
            else logger.error("Interaction ${interaction.name} not found")
        }
    }

    suspend fun listenVoiceState() {
        client.on<VoiceStateUpdateEvent> {
            runCatching { StateChange.voiceStateChange(this) }
                .onSuccess { logger.info("[VoiceStateChange] Executed successfully") }
                .onFailure { logger.error("[VoiceStateChange] ${it.message.toString()}") }
        }
    }

    suspend fun listenMessage() {
        client.on<MessageCreateEvent> {
            val member = member ?: return@on
            if (message.content == "/notify") FallbackNotifyCommand.sendReply(member, message)
        }
    }

    suspend fun login(vararg intent: Intent) {
        client.login {
            intents = Intents.nonPrivileged + Intents(*intent)
            presence {
                playing(Config.url)
            }
        }
    }

    companion object {
        suspend fun create(): Bot = Bot().init()
    }

}